<launch>
	<!-- Other options: retinanet, spnet, spanet -->
  <!-- See possible networks in https://github.com/personalrobotics/food_detector/ -->
  <!-- Set to "apriltag" for apriltag-based detection -->
  <arg name="detector" default="spanet" doc="Which network to use for food perception."/>
  <arg name="sim" default="false" doc="Whether to run in Aikido-sim only" />

  <!-- Simulated pose estimators -->
  <group if="$(arg sim)">
    <node name="st_sim_cantaloupe" pkg="tf" type="static_transform_publisher"
      respawn="false" output="screen" args="0.3 -0.25 0.232 0.0 0.0 1.0 0.0 map cantaloupe 10"/>
    <node name="st_sim_grape" pkg="tf" type="static_transform_publisher"
      respawn="false" output="screen" args="0.25 -0.29 0.232 0.0 0.0 1.0 0.0 map grape 10"/>

    <node pkg="ada_feeding"
          name="food_detector"
          type="run_sim_food_detector.py">
      <rosparam param="frame_list">["grape"]</rosparam>
    </node>

    <node pkg="ada_feeding"
          name="face_detector"
          type="run_sim_face_detector.py" />
  </group>

  <!-- Real Perception -->
  <group unless="$(arg sim)">
    <group unless="$(eval arg('detector') == 'apriltag')">
      <node pkg="food_detector" type="run_perception_module.py" name="food_detector"
            args="--demo-type $(arg detector)" output="screen" required="true"/>

      <node pkg="face_detection" type="face_detection" name="face_detection"
            required="true"/>
    </group>
    <group if="$(eval arg('detector') == 'apriltag')">
      <arg name="tag_image_topic" default="/camera/color/image_raw" doc="Image topic for apriltag" />
      <arg name="tag_camera_info" default="/camera/color/camera_info" doc="Camera Info topic for apriltag" />

      <arg name="camera_frame" default="camera_color_optical_frame" />
      <arg name="tag_frame" default="detected_food_tag" />
      
      <rosparam ns="apriltag_perception" file="$(find ada_feeding)/config/gen2_apriltag_perception.yaml" />

      <!-- Food Apriltag -->
      <node pkg="apriltag_ros" type="apriltag_ros_continuous_node" name="apriltag_perception" output="screen">
        <remap from="image_rect" to="$(arg tag_image_topic)"/>
        <remap from="camera_info" to="$(arg tag_camera_info)"/>
      </node>

      <!-- Marker Publisher -->
      <node pkg="ada_feeding" type="apriltag_food_perception.py" name="food_detector" output="screen">
        <remap from="image_compressed" to="$(arg tag_image_topic)/compressed"/>
        <remap from="camera_info" to="$(arg tag_camera_info)"/>
        <param name="camera_frame" value="$(arg camera_frame)" />
        <param name="tag_frame" value="$(arg tag_frame)" />
        <param name="publish_tag_detections_image" type="bool" value="false" />
      </node>
    </group>
  </group>
</launch>